### DISCUSSION TITLE
Error in FUN(X[[i]], ...) : object 'treatment' not found when add_pvalue to faceted plot

### DISCUSSION THREAD

**Discussion by jianwu1** [csdaw/ggprism#19] [Q&A]
Hi Charlotte @csdaw ,

Thank you very much for producing ggprism! It's such a fantastic tool for generating prism-style plots!

I produced a plot successfully before add_pvalue, but ran into this "`Error in FUN(X[[i]], ...) : object 'treatment' not found`" after trying to add pvalues. Most perplexingly, the error is caused by a variable "treatment", which is not involved in the p-value data frame for the "`test`" argument. It's a variable fed to the `aes()` in ggplot. I read the vignette multiple times but couldn't figure out why this error occurs. I also ran `debugonce(add_pvalue)` and the function ran ok. Can I ask for your help to figure out what the problem is please? 

This is the data and code I used:
[data.zip](https://github.com/csdaw/ggprism/files/9425832/data.zip)

Attached here is the error message:
<img width="366" alt="Screen Shot 2022-08-26 at 12 58 47 am" src="https://user-images.githubusercontent.com/62498686/186699800-a509015b-a909-4053-aed8-de1eea693b28.png">

Here is the plot that was successfully produced without add_pvalue():
![plot_without_p_value](https://user-images.githubusercontent.com/62498686/186700068-9f0099b8-c0e7-4f60-97bb-36dca90d7fcb.png)

Thank you very much!

Best wishes,

Jian

**Comment by csdaw** [comment]
Hi there,

Thanks for the code, I can reproduce the problem.

To just fix the error you need to change `aes(shape = treatment)` into `aes(group = treatment)`, and then separately put `aes(shape = treatment)` into `geom_jitter()`. 

Indeed you actually already found the answer to the problem! Using `ggplot(aes(shape = treatment))` will look for the column named `treatment` in every data frame referenced in the lines of code. But as you rightly point out, there is no column named `treatment` in the pvalue data frame `Foxp3_CD8.pvalue`.

This will fix the error, but then the brackets won't be plotting in the correct x position.

To fix this secondary issue, just remove the line `group = "treatment"` from `rstatix::add_xy_position()`. Lastly, you might like to adjust the bracket width a little with the `bracket.shorten` argument of `add_pvalue()`.

Hope this helps.

``` r
library(tidyverse)
library(ggprism)
library(rstatix)
#> 
#> Attaching package: 'rstatix'
#> The following object is masked from 'package:stats':
#> 
#>     filter

# load data
Foxp3_CD8 <- readRDS("~/Downloads/data/Foxp3_CD8.rds")


# create a stat dataframe with rstatix
Foxp3_CD8.pvalue <- Foxp3_CD8 %>% 
  rstatix::group_by(sample, value_type) %>% 
  rstatix::wilcox_test(value ~ treatment, detailed = TRUE) %>% 
  adjust_pvalue(method = "BH") %>% 
  add_significance(p.col = "p.adj") %>% 
  p_format("p.adj", digits = 2) %>% 
  rstatix::add_xy_position(
    x = "sample",
    #group = "treatment",
    fun = "mean_sd",
    scales = "free_y"
  )

# generate faceted plot with p-value
Foxp3_CD8 %>%   
  ggplot(aes(x=sample, y=value, group = treatment), color="black") +
  geom_jitter(
    aes(shape = treatment),
    size = 6,
    position = position_jitterdodge(dodge.width = 0.5)
  ) +
  stat_summary(
    geom = "bar",
    fun = mean,
    color = "black",
    position = position_dodge(width =0.5),
    fill = NA,
    size = 1.5,
    width = 0.5
  ) +
  stat_summary(
    geom = "errorbar",
    fun.min = function(.x){mean(.x) - sd(.x)},
    fun.max = function(.x){mean(.x) + sd(.x)},
    position = position_dodge(width =0.5),
    width = 0.2, 
  ) +
  scale_shape_manual(values = c(16, 1)) +
  facet_wrap(~ value_type, scales = "free") +
  theme_prism() +
  theme(axis.title.x = element_blank()) +
  add_pvalue(Foxp3_CD8.pvalue,
             xmin = "xmin",
             xmax = "xmax",
             label = "P = {p.adj}",
             bracket.shorten = 0.1
  )
```

![](https://i.imgur.com/5U6ahIn.png)

<sup>Created on 2022-08-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by jianwu1** [comment]
Hi Charlotte,

Thank you very much for your help!!! And thank you very much for your explanation about how ggplot aes arguments are inherited in add_xy_position()!

Best wishes,

Jian
